File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.py
'''
로봇 탐사하기

조건 : 한번 지나간 좌표는 다시 돌아갈 수 없다
    ==> 한 층에서는 일방통행만 가능하다는 뜻

    왼 -> 오
    오 -> 왼
    각 행 별 누적값 탐사해본 뒤, 두 방향의 최댓값 정보를 dp에 저장 (경로는 일방통행이므로 중복 방문하지 않는다)
'''

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]
dp = [[0] * m for _ in range(n)]

# 0. 첫번째 행 누적합 진행(init)
dp[0][0] = grid[0][0]
for i in range(1, m):
    dp[0][i] = dp[0][i - 1] + grid[0][i]

# 1. 각 행 돌면서 방향 별 누적합 구하기
for i in range(1, n):
    right, left = [0] * m, [0] * m

    right[0], left[-1] = dp[i - 1][0] + grid[i][0], dp[i - 1][-1] + grid[i][-1]
    # 1-1. 오른쪽 방향으로 이동
    for j in range(1, m):
        right[j] = max(dp[i - 1][j], right[j - 1]) + grid[i][j]
    # 1-2. 왼쪽 방향으로 이동
    for j in range(m - 2, -1, -1):
        left[j] = max(dp[i - 1][j], left[j + 1]) + grid[i][j]

    # 1-3. dp에 이동 간 max값 업데이트하기
    for j in range(m):
        dp[i][j] = max(right[j], left[j])

print(dp[-1][-1])

File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.cc
/*
왼쪽, 오른쪽, 아래쪽으로는 이동 가능
한 번 탐사한 지역을 다시 탐사하지는 않음

풀이:
DP
먼저 오른쪽으로 이동해가면서 더하기
그다음, 왼쪽으로 이동해가면서, 최댓값으로 업데이트해주기
	- 먼저 계산한다음 오른쪽에서 온 값과 비교
*/

#include <iostream>
#include <algorithm>
#include <vector>
#define INIT cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);

using namespace std;

int n, m;
int arr[1001][1001];
int dp[1001][1001] = { 0, };

int main() {
	cin >> n >> m;
	vector<int> right(m + 1);
	vector<int> left(m + 1);
	for (int i = 1; i < n + 1; i++) {
		for (int j = 1; j < m + 1; j++)
			cin >> arr[i][j];
	}

	// 0. 첫번째 줄 초기화
	dp[1][1] = arr[1][1];
	for (int i = 2; i < m + 1; i++)
		dp[1][i] = dp[1][i - 1] + arr[1][i];

	// 1. DP 시작
	for (int i = 2; i < n + 1; i++) {
		// 왼쪽 오른쪽 나눠서 계산한 뒤에 합치기
		right[1] = dp[i - 1][1] + arr[i][1];
		left[m] = dp[i - 1][m] + arr[i][m];

		// 2. 오른쪽부터 먼저 채우기
		for (int j = 2; j < m + 1; j++)
			right[j] = max(right[j - 1], dp[i - 1][j]) + arr[i][j];
		// 3. 왼쪽 값 채우기
		for (int j = m - 1; j > 0; j--)
			left[j] = max(left[j + 1], dp[i - 1][j]) + arr[i][j];
		// 4. dp테이블 최댓값으로 업데이트
		for (int j = 1; j < m + 1; j++)
			dp[i][j] = max(left[j], right[j]);

		// 5. left, right 초기화
		for (int i = 1; i < m + 1; i++) {
			right[i] = 0;
			left[i] = 0;
		}
	}

	cout << dp[n][m] << '\n';

	return 0;
}